R/clean_data_helpers.R

```diff
@@ -1,54 +1,145 @@
+  type    <- unlist(strsplit(type, ",", fixed = TRUE))
+  verdict <- type %in% c("logical", "factor")
+
+  ## logical and factor are processed differently from the other.
+  ## Logical vectors are expected to contain `TRUE` or `FALSE` or `NA`. For
+  ## columns of this type, we will check the %NA, %TRUE and %FALSE. The others
+  ## will be set at 0.
+  ## For factor columns, the processing is based on their levels. Levels will be
+  ## checked for numeric, date, logical, NA, and character
+  if (any(verdict)) {
+    return(scan_lgl_and_fct_columns(x))
+  }
+
+  # save the variable length
   n_rows <- length(x)
 
-  # --- get the proportion of NA ---
+  # get the proportion of NA
   are_na <- round((sum(is.na(x)) / n_rows), 6L)
   x      <- x[!is.na(x)]
 
-  # --- get the proportion of numeric values ---
-  tmp         <- suppressWarnings(as.numeric(x))
-  are_numeric <- round((sum(!is.na(tmp)) / n_rows), 6L)
-
-  # --- get the proportion of date values ---
-  x        <- x[which(is.na(tmp))]
+  # get the proportion of date values
   are_date <- 0L
+  if (any(type %in% c("POSIXct", "POSIXt"))) {
+    tmp      <- as.Date(x)
+    are_date <- round((sum(!is.na(tmp)) / n_rows), 6L)
+    x        <- x[is.na(tmp)]
+  }
   if (!is.null(lubridate::guess_formats(x, c("ymd", "ydm", "dmy", "mdy", "myd",
                                              "dym", "Ymd", "Ydm", "dmY", "mdY",
                                              "mYd", "dYm")))) {
-    x <- suppressWarnings(
+    tmp    <- suppressWarnings(
       as.Date(
         lubridate::parse_date_time(
           x, orders = c("ymd", "ydm", "dmy", "mdy", "myd", "dym", "Ymd", "Ydm",
                         "dmY", "mdY", "mYd", "dYm")
         )
       )
     )
-    are_date <- round((sum(!is.na(x)) / n_rows), 6L)
+    are_date <- round((sum(!is.na(tmp)) / n_rows), 6L)
+    x        <- x[is.na(tmp)]
   }
 
-  # --- get the proportion of logical values ---
-  are_logical   <- round((sum(is.logical(x)) / n_rows), 6L)
+  # get the proportion of numeric values
+  are_numeric <- are_logical <- 0L
+  if (length(x) > 0L) {
+    tmp         <- suppressWarnings(as.numeric(x))
+    are_numeric <- round((sum(!is.na(tmp)) / n_rows), 6L)
+  }
 
-  # --- get the proportion of character values ---
-  are_character <- round((1.0 - (are_na + are_numeric +
-                                   are_date + are_logical)), 6L)
+  # get the proportion of character values
+  are_character   <- 0L
+  if (!all(verdict)) {
+    are_character <- round((1.0 - (are_na + are_numeric +
+                                     are_date + are_logical)), 6L)
+  }
 
-  # --- return the output ---
+  # return the output
   return(c(are_na, are_numeric, are_date, are_character, are_logical))
 }
 
+#' Get class of a vector
+#'
+#' @param x The input vector
+#'
+#' @return A character with the class of the input vector
+#' @keywords internal
+#'
+get_class <- function(x) {
+  paste(class(x), collapse = ",")
+}
```

Comment:
In general, it's a bit idea to manipulate objects / code as text, especially to circumvent explicit safeguards